heirloom/grep/grep.1
'\" t
.\" Sccsid @(#)grep.1	1.30 (gritter) 3/6/04
.\" Parts taken from grep(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"   Redistributions of source code and documentation must retain the
.\"    above copyright notice, this list of conditions and the following
.\"    disclaimer.
.\"   Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"   All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\"   Neither the name of Caldera International, Inc. nor the names of
.\"    other contributors may be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE
.\" LIABLE FOR ANY DIRECT, INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.TH GREP 1 "3/6/04" "Heirloom Toolchest" "User Commands"
.SH NAME
grep \- search a file for a pattern
.SH SYNOPSIS
.HP
.ad l
.nh
\fB/usr/5bin/grep\fR [\fB\-bchilnsv\fR]
\fIpattern\fR [\fIfile\fR\ ...]
.HP
.PD 0
.ad l
\fB/usr/5bin/posix/grep\fR [\fB\-E\fR|\fB\-F\fR]
\fB\-e\fI\ pattern_list\fR\ ...
[\fB\-f\fI\ pattern_file\fR] [\fB\-c\fR|\fB\-l\fR|\fB\-q\fR]
[\fB\-bhinsvx\fR] [\fIfile\fR\ ...]
.HP
.ad l
\fB/usr/5bin/posix/grep\fR [\fB\-E\fR|\fB\-F\fR]
\fB\-f\fI\ pattern_file\fR
[\fB\-e\fI\ pattern_list\fR\ ...] [\fB\-c\fR|\fB\-l\fR|\fB\-q\fR]
[\fB\-bhinsvx\fR] [\fIfile\fR\ ...]
.HP
.ad l
\fB/usr/5bin/posix/grep\fR [\fB\-E\fR|\fB\-F\fR]
[\fB\-c\fR|\fB\-l\fR|\fB\-q\fR] [\fB\-bhinsvx\fR]
\fIpattern_list\fR [\fIfile\fR\ ...]
.br
.PD
.ad b
.hy 1
.SH DESCRIPTION
The
.B grep
command searches the lines of the specified files
(or of standard input)
for occurrences of the regular expression
.I pattern.
The default behavior is to print each matching line to standard output.
.PP
The
.B /usr/5bin/grep
command accepts one pattern
that is treated as a simple regular expression;
it uses a compact nondeterministic algorithm.
.PP
The
.B /usr/5bin/posix/grep
command uses basic regular expressions by default
and accepts a newline-separated list of patterns
as described for the
.B \-e
option below.
It uses a deterministic algorithm with moderate space requirements
for most expressions;
backreferences, word delimiters, and multi-character collating elements
cause a nondeterministic algorithm to be used.
.PP
.B /usr/5bin/s42/grep
is identical to
.BR /usr/5bin/posix/grep .
.PP
See the description of
.IR ed (1)
for the specifications of simple and basic regular expressions.
.PP
Care should be taken when using the characters
$ * [ ^ | ? ' " ( ) and \e in the expression
as they are also meaningful to the Shell.
It is safest to enclose the entire expression
argument in single quotes ' '.
.PP
Both
.B /usr/5bin/grep
and
.B /usr/5bin/posix/grep
accept the following options:
.TP
.B \-b
Each line is preceded by the block number on which it was found.
This is sometimes useful
in locating disk block numbers by context.
Block numbers start with 0.
.TP
.B \-c
Only a count of matching lines is printed.
.TP
.B \-h
Normally, the name of each input file is printed before a match
if there is more that one input file.
When this option is present, no file names are printed.
.TP
.B \-i
Upper- and lowercase differences are ignored when searching matches.
.TP
.B \-l
The names of files with matching lines are listed
(once) separated by newlines.
.TP
.B \-n
Each line is preceded by its line number in the file.
Line numbers start with 1.
.TP
.B \-s
Error messages for nonexistent or unreadable files are suppressed.
.TP
.B \-v
All lines but those matching are printed.
.SS /usr/5bin/posix/grep
The following options are supported by
.B /usr/5bin/posix/grep
only:
.TP
.BI \-e\  pattern_list
Specifies one or more patterns, separated by newline characters.
A line is selected if one or more of the specified patterns are found.
.TP
.B \-E
All patterns are interpreted as extended regular expressions
as described in
.IR egrep (1).
.TP
.BI \-f\  pattern_file
One or more patterns, separated by newline
characters, are read from
.I pattern_file.
.TP
.B \-F
All patterns are interpreted as fixed strings,
as with
.IR fgrep (1).
.TP
.B \-q
Do not write anything to standard output.
.TP
.B \-x
Consider only lines consisting of the pattern as a whole,
like a regular expression surrounded by
.I ^
and
.I $.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_COLLATE
Affects the collation order for range expressions,
equivalence classes, and collation symbols
in basic regular expressions.
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters
in both simple and basic regular expressions,
the availability and composition of character classes
in basic regular expressions,
and the case mapping for the
.B \-i
option.
.SH "SEE ALSO"
ed(1),
egrep(1),
fgrep(1),
sed(1),
locale(7)
.SH DIAGNOSTICS
Exit status is 0 if any matches are found,
1 if none, 2 for syntax errors or inaccessible files.
.SH NOTES
If a line contains a
.SM NUL
character,
only matches up to this character are found
(unless
.B /usr/5bin/posix/grep
is used with the
.I \-F
option).
The entire matching line will be printed.
.PP
The LC_COLLATE variable has currently no effect.
Ranges in bracket expressions are ordered
as byte values in single-byte locales
and as wide character values in multibyte locales;
equivalence classes match the given character only,
and multi-character collating elements are not available.
.PP
The options supported by
.B /usr/5bin/posix/grep
that are not accepted by
.B /usr/5bin/grep
can easily be replaced by portable constructs:
Use
.I egrep
instead of
.BR \-E ,
.I fgrep
instead of
.BR \-F .
Use
.I egrep
if you need the
.B \-e
or
.B \-f
option,
use only one of them and that only once;
if necessary, use text processing tools
to generate a single expression list before.
Redirect standard output to
.I /dev/null
for
.B \-q
(the possible speedup with
.I \-q
is never worth human time spent with porting scripts),
and use the
.RI ` ^ '
and
.RI ` $ '
meta-characters instead of
.BR \-x .


ex-vi/Changes
Release ...
* The screen buffers for visual mode are now dynamically allocated, so
  vi usually does not return to ex mode with "screen too large" when the
  terminal is resized on a large monitor anymore.
* ex can now edit files with lines of arbitrary length. vi currently only
  shows the beginning of a line that does not fit onto the screen in its
  entirety.
* Viewing executables and compressed files is no longer inhibited.

Release 3/25/05
* vi no longer dies with a segmentation fault if a line does not fit on the
  screen after an insertion.
* The 'p' command now works correctly if the buffer contains a partial line
  with multibyte characters.
* Traditional regular expressions sometimes failed to operate correctly
  since the last release.

Release 2/25/05
* Traditional regular expressions can now be used with multibyte characters.
* When the 'ignorecase' option is toggled, saved regular expressions are now
  updated accordingly. (P)
* If a line began with a tabulator and another tabulator was inserted with
  the cursor located on the first tabulator, the display was not updated
  appropriately since the last release (Bugreport by Matthew Fischer). (P)
* Fixed a segmentation fault that occured in multibyte locales when operation
  was continued after vi had been stopped by ^Z, with the cursor positioned
  in a line longer than the terminal width as the last line of the current
  screen.
* Made multicolumn characters work in lines longer than the screen width in
  visual mode (Bugreport by Matthew Fischer).
* Made it work for Big5 locales (Patches by Matthew Fischer).
* Fixed a problem with the 'r' command in EUC-JP and Big5 locales (Bugreport
  by Matthew Fischer).
* The insertion of multicolumn characters and tab characters in multibyte
  locales now works with terminals that have the 'ic' but no 'im' termcap
  capability (Bugreport by Matthew Fischer).
* The argument to the -w option is correctly recognized now.
* If the SHELL environment variable is set to the empty string, it is now
  ignored.
* A non-null exit status is now returned if a file could not be opened, if
  an invalid address is given for a command, or if a tag is not found.
* If the match for a substitution is of length zero, a line overflow is
  now avoided when compiled with the 'UNIX(R) Regular Expression Library'.
* When moving left while the cursor is positioned over a multicolumn
  character at the end of the line, the bell is rung now (Bugreport by
  Matthew Fischer).
* When moving up or down to a row with different column arrangement while
  the cursor is positioned over a multicolumn character, the leftmost
  character above the original position is chosen in the new row.
* If both the -t and the -c option are given, the -t option is now
  processed first, i.e. the command is executed at the position where
  the tag was found.
* The -w option now also sets the scroll size for the 'z' command.
* When the name of a nonexisting file is given with the 'edit' ex command,
  the command now succeeds and prints a "[New file]" message.
* If standard output is not a terminal, no '\r' is written at the end of
  printed lines anymore.
* The 'source' ex command now works if command input comes from a pipe or
  regular file.
* Ex does not exit on errors immediately anymore if standard input is not
  a terminal but a pipe or regular file.
* The 'substitute' ex command can now be abbreviated as 'sub', 'subst' etc.
* A new version of mapmalloc.c that is derived from Unix 7th Edition code
  has been introduced.
* If the 'next!' ex command is given and the autowrite option is set, the
  current file is not written anymore.

Release 1/19/05
* The last release erroneously made 'X' work like 'x' in visual mode. It now
  deletes the character before the cursor again as documented (Bugreport by
  Matthew Fischer). (P)
* When a multicolumn character was replaced by another multicolumn character
  in insert mode, the display was not updated appropriately with terminals
  other than xterm.
* Various rendering errors happened with multicolumn characters if they
  started at an even column (counting from 1 upwards).
* When a multicolumn character was inserted and then replaced, the visual
  screen representation was sometimes not updated accordingly.
* Undoing the replacement of a multicolumn character by a singlecolumn
  character twice made the singlecolumn character invisible.
* The 'cw' visual command with a multibyte character as last character of
  the affected word located at the end of the line left garbage bytes past
  the end of the new text.
* Visual 'U' followed by 'u' lead to garbage on the screen when multibyte
  characters were present on the changed line.
* The position of '$' when changing text was fixed for cases where the first
  changed character had multiple columns but the last one had not.
* The handling of multicolumn characters was fixed for terminals without the
  IM (insert mode) capability. It is unlikely that such terminals actually
  exist, but vi will use the corresponding code for open mode if a termcap
  entry is not available.
* When an illegal multibyte sequence is entered in vi insert mode, no garbage
  bytes are generated anymore when the insert mode is left.
* The '\u', '\l', '\U', and '\L' substitution sequences work with multibyte
  characters now.
* Handle character case conversions with the '~' vi command correctly if the
  length of the converted multibyte sequence is smaller than the original one.
* Multibyte sequences that correspond to an unprintable character are now
  printed as multiple octal escape sequences.
* Pressing the ^U (kill) character in insert mode with a multibyte character
  at the beginning of an insertion at the start of a line sometimes positioned
  the cursor at weird locations since the last revision.
* Fixed an old vi bug: If a vi command that yanked or deleted part of a line
  was followed by an ex command that also yanked or deleted some text, a

heirloom/grep/grep.c
/*
 * grep - search a file for a pattern
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, April 2001.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

/*	Sccsid @(#)grep.c	1.44 (gritter) 12/2/04>	*/

/*
 * Code common to all grep flavors.
 */

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<sys/mman.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<libgen.h>
#include	<locale.h>
#include	<limits.h>
#include	<ctype.h>
#include	<dirent.h>

#include	"grep.h"
#include	"alloc.h"

/*
 * Generic flags and the like.
 */
int		Eflag;			/* use EREs */
int		Fflag;			/* use fixed strings */
int		bflag;			/* print buffer count */
int		cflag;			/* print count only */
int		fflag;			/* had pattern file argument */
int		hflag;			/* do not print filenames */
int		iflag;			/* ignore case */
int		lflag;			/* print filenames only */
int		nflag;			/* print line numbers */
int		qflag;			/* no output at all */
int		rflag;			/* operate recursively */
int		sflag;			/* avoid error messages */
int		vflag;			/* inverse selection */
int		xflag;			/* match entire line */
int		mb_cur_max;		/* avoid multiple calls to MB_CUR_MAX */
unsigned	status = 1;		/* exit status */
off_t		lmatch;			/* count of line matches */
off_t		lineno;			/* current line number */
char		*progname;		/* argv[0] to main() */
char		*filename;		/* name of current file */
char		*options;		/* for getopt() */
void		(*build)(void);		/* compile function */
int		(*match)(const char *, size_t); /* comparison function */
int		(*range)(struct iblok *, char *); /* grep range of lines */

/*
 * Regexp variables.
 */
struct expr	*e0;			/* start of expression list */
enum matchflags	matchflags;		/* matcher flags */

/*
 * Lower-case a character string.
 */
void
loconv(register char *dst, register char *src, size_t sz)
{
	if (mbcode) {
		char	mb[MB_LEN_MAX];
		wchar_t wc;
		int len, i;

		while (sz > 0) {
			if ((*src & 0200) == 0) {
				*dst++ = tolower(*src);
				src++;
				sz--;
			} else if ((len = mbtowc(&wc, src, sz)) <= 0 ||
					len > sz) {
				*dst++ = *src++;
				sz--;
			} else {
				wc = towlower(wc);
				if (len >= mb_cur_max) {
					if (wctomb(dst, wc) == len) {
						dst += len;
						src += len;
						sz -= len;
					} else {
						*dst++ = *src++;
						sz--;
					}
				} else {
					if (wctomb(mb, wc) == len) {
						sz -= len;
						src += len;
						for (i = 0; i < len; i++)
							*dst++ = mb[i];
					} else {
						*dst++ = *src++;
						sz--;
					}
				}
			}
		}
	} else
		while (sz--) {
			*dst++ = tolower(*src & 0377);
			src++;
		}
}

/*
 * Report a matching line.
 */
void
report(const char *line, size_t llen, off_t bcnt, int addnl)
{
	if (filename && !hflag)
		printf("%s:", filename);
#ifdef	LONGLONG
	if (bflag)
		printf("%llu:", (long long)bcnt);
	if (nflag)
		printf("%llu:", (long long)lineno);
#else	/* !LONGLONG */
	if (bflag)
		printf("%lu:", (long)bcnt);
	if (nflag)
		printf("%lu:", (long)lineno);
#endif	/* !LONGLONG */
	if (line && llen)
		fwrite(line, sizeof *line, llen, stdout);
	if (addnl)
		putchar('\n');
}

/*
 * Check line for match. If necessary, the line gets NUL-terminated (so
 * its address range must be writable then). When ignoring character case,
 * a lower-case-only copy of the line is made instead. If a match is found,
 * statistics are printed. Returns 1 if main loop shall terminate, 0 else.
 */
static int
matchline(char *line, size_t sz, int putnl, struct iblok *ip)
{
	int terminate = 0;
	char lbuf[512], *abuf = NULL, *cline = line;

	if (iflag && (matchflags & MF_LOCONV)) {
		if (sz >= sizeof lbuf - 1) {
			abuf = smalloc(sz + 1);
			cline = abuf;
		} else
			cline = lbuf;
		loconv(cline, line, sz);
		cline[sz] = '\0';
	} else if (matchflags & MF_NULTERM)
		cline[sz] = '\0';
	lineno++;
	if (match(cline, sz) ^ vflag) {
		lmatch++;
		if (qflag == 0) {
			if (status == 1)
				status = 0;
			if (lflag) {
				puts(filename ? filename : stdinmsg);
			} else if (!cflag)
				report(line, sz, (ib_offs(ip)-1) / BSZ, putnl);
		} else
			exit(0);
		if (qflag || lflag)
			terminate = 1;
	}
	if (abuf)
		free(abuf);
	return terminate;
}

/*
 * Check all lines within ip->ib_cur and last which contains the last
 * newline. If the main loop shall terminate, 1 is returned.
 */
static int
gn_range(struct iblok *ip, char *last)
{
	char *nl;

	while ((nl = memchr(ip->ib_cur, '\n', last + 1 - ip->ib_cur)) != NULL) {
		if (matchline(ip->ib_cur, nl - ip->ib_cur,  1, ip))
			return 1;
		if (nl == last)
			return 0;
		ip->ib_cur = nl + 1;
	}
	return 0;
}

/*
 * Main grep routine. The line buffer herein is only used for overlaps
 * between file buffer fills.
 */
static int
grep(struct iblok *ip)
{
	char *line = NULL;		/* line buffer */
	register char *lastnl;		/* last newline in file buffer */
	size_t sz = 0;			/* length of line in line buffer */
	char *cp;
	int hadnl;			/* lastnl points to newline char */
	int	oom = 0;		/* got out of memory */

	lineno = lmatch = 0;
	if (ib_read(ip) == EOF)
		goto endgrep;
	ip->ib_cur--;
	for (;;) {
		for (lastnl = ip->ib_end - 1;
				*lastnl != '\n' && lastnl > ip->ib_cur;
				lastnl--);
		if (hadnl = (ip->ib_cur < ip->ib_end && *lastnl == '\n'))
			if (range(ip, lastnl))
				break;
		if (lastnl < ip->ib_end - 1) {
			/*
			 * Copy the partial line from file buffer to line
			 * buffer. Allocate enough space to zero-terminate
			 * the line later if necessary.
			 */
			sz = ip->ib_end - lastnl - hadnl;
			line = smalloc(sz + 1);
			memcpy(line, lastnl + hadnl, sz);
			ip->ib_cur = lastnl + hadnl;
		} else
			line = NULL;
nextbuf:
		if (ib_read(ip) == EOF) {
			if (line) {
				matchline(line, sz, sus, ip);
				free(line);
				line = NULL;
				sz = 0;
			}
			break;
		}
		ip->ib_cur--;
		if (line) {
			/*
			 * Append the partial line at the beginning of the
			 * file buffer to the line buffer.
			 */
			size_t oldsz = sz;
			if ((cp = memchr(ip->ib_cur, '\n',
					ip->ib_end - ip->ib_cur)) == NULL) {
				char	*nline;
				/*
				 * Ugh. This is really a huge line. Store the
				 * entire file buffer in the line buffer and
				 * read the next part of the file.
				 */
				sz += ip->ib_end - ip->ib_cur;
				if ((nline = realloc(line, sz + 1)) == NULL) {
					sz = oldsz;
					cp = &ip->ib_end[-1];
					oom++;
				} else {
					line = nline;
					memcpy(line + oldsz, ip->ib_cur,
						ip->ib_end - ip->ib_cur);
					goto nextbuf;
				}
			}
			if ((sz = cp - ip->ib_cur) > 0) {
				char	*nline;
				sz += oldsz;
				if ((nline = realloc(line, sz + 1)) == NULL) {
					sz = oldsz;
					oom++;
				} else {
					line = nline;
					memcpy(line + oldsz, ip->ib_cur,
							cp - ip->ib_cur);
				}
			} else
				sz = oldsz;
			if (matchline(line, sz, 1, ip))
				break;
			free(line);
			line = NULL;
			sz = 0;
			ip->ib_cur = cp + (oom == 0);
			oom = 0;
		}
	}
endgrep:
	if (!qflag && cflag) {
		if (filename && !hflag)
			printf("%s:", filename);
#ifdef	LONGLONG
		printf("%llu\n", (long long)lmatch);
#else
		printf("%lu\n", (long)lmatch);
#endif
	}
	return 0;
}

/*
 * Grep a named file.
 */
static void
fngrep(const char *fn, int level)
{
	struct iblok	*ip;
#ifdef	ADDONS
	struct stat	st;

	if (rflag && fn && stat(fn, &st) == 0) {
		switch (st.st_mode&S_IFMT) {
#define	ignoring(t, s)	fprintf(stderr, "%s: ignoring %s %s\n", progname, t, s)
		case S_IFIFO:
			ignoring("named pipe", fn);
			return;
		case S_IFBLK:
			ignoring("block device", fn);
			return;
		case S_IFCHR:
			ignoring("block device", fn);
			return;
#ifdef	S_IFSOCK
		case S_IFSOCK:
			ignoring("socket", fn);
			return;
#endif	/* S_IFSOCK */
		default:
			break;
		case S_IFDIR: {
			char	*path;
			int	pend, psize, i;
			DIR	*df;
			struct dirent	*dp;

			if (hflag == 2)
				hflag = 0;
			if ((df = opendir(fn)) == NULL) {
				if (sflag == 0)
					fprintf(stderr, "%s: can't open "
							"directory %s\n",
							progname, fn);
				if (!qflag || status == 1)
					status = 2;
				return;
			}
			pend = strlen(fn);
			path = malloc(psize = pend + 2);
			strcpy(path, fn);
			path[pend++] = '/';
			while ((dp = readdir(df)) != NULL) {
				if (dp->d_name[0] == '.' &&
						(dp->d_name[1] == '\0' ||
					 	dp->d_name[1] == '.' &&
					 	dp->d_name[2] == '\0'))
					continue;
				i = 0;
				do {
					if (pend + i >= psize)
						path = srealloc(path,
								psize += 14);
					path[pend+i] = dp->d_name[i];
				} while (dp->d_name[i++]);
				filename = path;
				fngrep(path, level+1);
			}
			free(path);
			closedir(df);
			return;
		    }
		}
	}
#endif	/* ADDONS */
	if (fn) {
		if ((ip = ib_open(fn, 0)) == NULL) {
			if (sflag == 0)
				fprintf(stderr, "%s: can't open %s\n",
						progname, fn);
			if (!qflag || status == 1)
				status = 2;
			return;
		}
	} else
		ip = ib_alloc(0, 0);
	grep(ip);
	if (ip->ib_fd)
		ib_close(ip);
	else
		ib_free(ip);
}

int
main(int argc, char **argv)
{
	int i, hadpat = 0;

#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	progname = basename(argv[0]);
	setlocale(LC_CTYPE, "");
	setlocale(LC_COLLATE, "");
	mb_cur_max = MB_CUR_MAX;
	range = gn_range;
	init();
	while ((i = getopt(argc, argv, options)) != EOF) {
		switch (i) {
		case 'E':
			Eflag |= 1;
			rc_select();
			break;
		case 'F':
			if (Eflag&2)
				Eflag = 0;
			Fflag |= 1;
			ac_select();
			break;
		case 'b':
			bflag = 1;
			break;
		case 'c':
			cflag = 1;
			break;
		case 'e':
			patstring(optarg);
			hadpat++;
			break;
		case 'f':
			fflag++;
			patfile(optarg);
			hadpat++;
			break;
		case 'h':
			hflag = 1;
			break;
		case 'i':
		case 'y':
			iflag = 1;
			break;
		case 'l':
			lflag = 1;
			break;
		case 'n':
			nflag = 1;
			break;
		case 'q':
			qflag = 1;
			break;
#ifdef	ADDONS
		case 'r':
			rflag = 1;
			break;
#endif	/* ADDONS */
		case 's':
			sflag = 1;
			break;
		case 'v':
			vflag = 1;
			break;
		case 'x':
			xflag = 1;
			break;
		default:
			if (!(Fflag&2))
				usage();
			status = 2;
		}
	}
	if (sus) {
		if (Fflag == 2) {
			if (sflag) {
				optind = 1;
				argv[1] = "-s";
				getopt(argc, argv, "");
				usage();
			}
			if (qflag) {
				optind = 1;
				argv[1] = "-q";
				getopt(argc, argv, "");
				usage();
			}
		}
		if (Fflag && status == 2)
			usage();
		if (Eflag == 1 && Fflag == 1 || cflag + lflag + qflag > 1)
			usage();
	}
	if (cflag)
		lflag = 0;
	if (hadpat == 0) {
		if (optind >= argc)
			misop();
		patstring(argv[optind++]);
	} else if (e0 == NULL)
		patstring(NULL);
	build();
	if (optind != argc) {
		if (optind + 1 == argc)
			hflag = 2;
		do {
			if (sus && argv[optind][0] == '-' &&
					argv[optind][1] == '\0') {
				filename = NULL;
				fngrep(NULL, 0);
			} else {
				filename = argv[optind];
				fngrep(argv[optind], 0);
			}
		} while (++optind < argc);
	} else {
		if (lflag && !sus && (Eflag || Fflag))
			exit(1);
		fngrep(NULL, 0);
	}
	return status;
}
